- Description: zfcp: Error recovery does not finish after a cable to Shark is unplugged
  Symptom:     When a cable to a Shark is unplugged/plugged error recovery would never being completed
  Problem:     If an ELS request is not processed by adapter ZFCP doesn't do any action to recover a link
  Solution:    Force reopen of port if link test procedure failed
  Problem-ID:  9824
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/scsi/zfcp_erp.c	1 Jul 2004 09:01:45 -0000	1.49.2.5
+++ linux-2.5/drivers/s390/scsi/zfcp_erp.c	5 Jul 2004 13:39:19 -0000	1.49.2.6
@@ -31,7 +31,7 @@
 #define ZFCP_LOG_AREA			ZFCP_LOG_AREA_ERP
 
 /* this drivers version (do not edit !!! generated and updated by cvs) */
-#define ZFCP_ERP_REVISION "$Revision: 1.49.2.5 $"
+#define ZFCP_ERP_REVISION "$Revision: 1.49.2.6 $"
 
 #include "zfcp_ext.h"
 
@@ -435,8 +435,20 @@
 	u8 req_code, resp_code;
 	int retval = 0;
 
-	if (send_els->status != 0)
+	if (send_els->status != 0) {
+		ZFCP_LOG_NORMAL("ELS request timed out, force physical port "
+				"reopen of port 0x%016Lx on adapter %s\n",
+				port->wwpn, zfcp_get_busid_by_port(port));
+		debug_text_event(port->adapter->erp_dbf, 3, "forcreop");
+		retval = zfcp_erp_port_forced_reopen(port, 0);
+		if (retval != 0) {
+			ZFCP_LOG_NORMAL("reopen of remote port 0x%016Lx "
+					"on adapter %s failed\n", port->wwpn,
+					zfcp_get_busid_by_port(port));
+			retval = -EPERM;
+		}
 		goto skip_fsfstatus;
+	}
 
 	req = (void*)((page_to_pfn(send_els->req->page) << PAGE_SHIFT) + send_els->req->offset);
 	resp = (void*)((page_to_pfn(send_els->resp->page) << PAGE_SHIFT) + send_els->resp->offset);

--- linux-2.5/drivers/s390/scsi/zfcp_def.h	15 Jun 2004 08:33:44 -0000	1.71.2.3
+++ linux-2.5/drivers/s390/scsi/zfcp_def.h	5 Jul 2004 13:39:19 -0000	1.71.2.4
@@ -33,7 +33,7 @@
 #define ZFCP_DEF_H
 
 /* this drivers version (do not edit !!! generated and updated by cvs) */
-#define ZFCP_DEF_REVISION "$Revision: 1.71.2.3 $"
+#define ZFCP_DEF_REVISION "$Revision: 1.71.2.4 $"
 
 /*************************** INCLUDES *****************************************/
 
@@ -68,7 +68,7 @@
 /********************* GENERAL DEFINES *********************************/
 
 /* zfcp version number, it consists of major, minor, and patch-level number */
-#define ZFCP_VERSION		"4.0.0"
+#define ZFCP_VERSION		"4.0.1"
 
 static inline void *
 zfcp_sg_to_address(struct scatterlist *list)

